### ISSUE TITLE
Problem getting Cross-Toolchain running

### ISSUE THREAD

**Issue by zabeloliver** [conan-io/conan#1539]
Hi, i'm trying to setup a reliable build environment where all needed tools and libs are fetched from conan. My projects consists of embedded micro-controller targets cross-build under windows.
So far i have created a toolchain-package which contains the toolchain binaries. 

Here is the conanfile.py
 ``` python
def source(self):
        # tools.download("some windows url for 4.8", "zipname.zip")
        self.run("git clone ssh://git@.../test/cypress-arm-gcc-49.git")
        self.run("cd cypress-arm-gcc-49 && git checkout v4.9.3")

    def package(self):
        self.copy(pattern="*", src="cypress-arm-gcc-49", dst="toolchain", keep_path=True)

    def package_info(self):
        self.env_info.CC =  os.path.join(self.package_folder, "toolchain", "bin", "arm-linux-gnueabi-gcc.exe")
        self.env_info.CXX = os.path.join(self.package_folder, "toolchain", "bin", "arm-linux-gnueabi-g++.exe")
``` 

No i' struggeling with the next step. I want to use my existing croos toolchain file for CMake to point to the compiler . But there is no ENV var set after the install cmd. Is there something i miss?

My complete approach would be (pseudo code):
```
conan install .. --build missing
cmake -DCMAK_TOOLCHAIN_FILE=...
cmake --build . --config Debug
```
Is this correct?
Since i'm new to conan all comments are helpful :)

Thanks in advance!

Best regards, 
Oli

**Comment by lasote** [comment]
Hi! it sounds very interesting.
So, you have the conan toolchain package but then you want to build by yourself your project, right?

In this case, you will need the [virtualenvironments](http://docs.conan.io/en/latest/mastering/virtualenv.html). If you specify `virtualenv` as a generator in your conan install (with -g or if you use a consumer conanfile, specify in [generators] section). It will create in the current directory a couple of scripts  (activate.bat and deactivate.bat) that will set the CC and CXX environment variables available. Then you can call your toolchain.

If you are going to build other conan packages using that toolchain conan package, you can specify a ["build_require"](http://docs.conan.io/en/latest/mastering/build_requires.html) in your packages with the toolchain, that way, the environment variables will be applied automatically and will be available in your build method automatically.

Thanks!

**Comment by memsharded** [comment]
A few extra suggestions, just in case some is useful:

- build missing
 ``conan install .. --build missing``
If you are creating a package for a toolchain, that will always be retrieved, you can use the ``build_policy="missing"`` in the package recipe, to specify that, so users will just do ``conan install ..``

- cmake calls
```
cmake -DCMAK_TOOLCHAIN_FILE=...
cmake --build . --config Debug
```
If you are using a gcc compiler, it is possible that you are using Makefiles as build system, right? In that case, you want to use the following cmake calls:
```
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE=...
cmake --build .
```

- CMake toolchain file
What is your plan to distribute and use this file? You might be able to package it with the toolchain itself, then ``imports`` it, so you get a copy in your folder after install, ready to be used.

**Comment by zabeloliver** [comment]
Hi, 
first of all: THANK YOU BOTH! 
@lasote The tip with the virtualenv is working - the project now builds with the packed toolchain. :+1: 
@memsharded Thanks for your input. Reagrding distribution of the toolchain-file - i will try the `imports` feature. Sounds like a good idea to distribute it together with the toolchain.

I think, i'm now on a good way, to pack all needed tools and setup a reliable toolchain! Our company needs this due to legal reasons. So this was important to get this running!

Thanks again!

**Comment by memsharded** [comment]
You are welcome, glad to help :)

If this is working, I think we can close the issue. But please re-open it if you have further related questions, or open a new issue for new ones. Also, if you need something that you don't want to discuss in public, due to those legal constraints, you can also send us email too.

Cheers!

**Comment by zabeloliver** [comment]
Hi, i have to reopen this issue since i need a little bit more help.
I have now toolchain packages which defines "virtualenv" as generator and i export a cmake toolchain file. In a cmake test project i need to call:
```
conan install ..
activate.bat
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE=...
cmake --build .
```
This is working. Now i want to build packages on which uses this toolchain package. ATM it looks like this:

```python
class MyModule(ConanFile):
    name = "module"
    version = "0.1"
    settings = "os", "compiler", "arch"
    exports_sources = "src/*"
    requires = "mingw-make/3.82@user/testing"
    
    def imports(self):
        self.copy("*.cmake", "", "")
        self.copy("*.bat",  "", "")

    def build(self):        
        if self.settings.os == "none" and self.settings.compiler == "armgcc":
            if self.settings.compiler.version == "4.9.3":
                self.build_requires("arm-gcc/4.9.3@user/stable")
                cmake = CMake(self)
                self.run('cmake %s/src -G"MinGW Makefiles"' % (self.source_folder))
                self.run("cmake --build . %s" % cmake.build_config)
            elif self.settings.compiler.version == "6.2":
                self.build_requires("arm-gcc/6.2@user/stable")
            else:
                return
        
    def package(self):
        self.copy("*.h", dst="include", keep_path=False)
        self.copy("*.a", dst="lib", keep_path=False)
    
    def package_info(self):
        self.cpp_info.libs = ["module"]
```
But it does not find the toolchain file which i somehow need to import and somehow activate the virtualenv that cmake finds the toolchain. Can you please give me a short hint howto do this?

Thanks in advance!
Best regards, 
Oliver

**Comment by memsharded** [comment]
The generators are something that are declared by the package that consumes dependencies, not by the dependencies themselves. In fact, they always generate information file for the dependencies exclusively.

In conanfile.py you don't need to use such generators, as the environment variables of the dependencies are already activated!

Also, the ``self.build_requires`` inside the ``build()`` method is not valid, it has to be at the class level, or in the ``build_requirements()`` method.

Probably you also want ``requires = "mingw-make/3.82@user/testing"`` to be a ``build_requires``

It is a bit confusing which package you are creating. The first lines doesn't create any package at all, they just install existing packages and consume them. So I would need to know which package are you creating, how do you create it, then a very simple consumer example of just that package (which by the way, could be well in ``test_package`` to automate it). Please tell me a bit more. If you would like to chat, there is an unofficial "conan" channel in the cpplang slack team, if you want to join.

Thanks!

**Comment by zabeloliver** [comment]
hi, 
ok, sorry, i was a bit in a hurry while writing the comment. Let's explain a bit what my goal is.
I'm working for a company in the white goods industry where we deliver electronic components based on various embedded platforms such as ARM, Mircochip, Renesas and so on. So many different, most of the time, non standard compilers. Since we are in charge in case of any problems in field we need to be able to build the excat same binary of a released software even after 5 to 10 years. So our current approach is to store the toolchain which is used to build the binary aside the code in our SVN server. So we can checkout out each historical revision and build it. This approach has many downsides, for example having two same checked-out toolchains in two projects - because it is added as svn:external. Another downside is, that many modules are included as externals - which is in man point of view a pain in th a**. 

So now, i'm trying to bring the software development a bit out of the past and want to implement conan as solution for our needs. That means, having all build tools needed for building the binary as packages in a conan server.

So i started to play a bit with a little test project based on an arm gcc compiler on windows. First i moved the building project from an IDE build to a cmake project. At this step i learned, that it will be necessary to have each toolchain version in package beside a toolchain file. So i managed to pack the gcc toolchain with the toolchain file into a package and noticed, that it seems to be a good way that the toolchains have a virtualenv generator to avoid path problems. 

That worked, and this is what i meant with 
```
conan install ..
activate.bat
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE=...
cmake --build .
´´´
I installed the toolchain, imported the toolchain file, activated the environment, started the build. This is working. 

Now i started to get rid of my svn:externals by building modules, which are build by the toolchains in the toolchain packages. 

So basicly somehting like
```
conan export user/stable -s compiler=armgcc -s compiler.version=4.9 
or
conan export user/stable -s compiler=rxc -s compiler.version=1.00.2.09
```

Therefor i want to create a conanfile.py which basicly does what i tried with my test project.
Get the toolchain, activate the virtualenv, build and upload the package. This is what i tried with snippet of my conanfile above. I created a little module which should be build as a library. A .h, a .c and a .cmake file - similar to the last webinar "whats new in conan".

With this approach i would be able to produce modules/libs which are compiled with a specific toolchain version for a specific platform - even after years.
As bonus i want to be able to build it with a mingw gcc on windows to add unit tests.

Unfortunately the amount of possibilities of conan does not make it easy for me to figure out how to solve my problem. I'm pretty sure, that conan is capable of achiving exactly this task.

Sorry for my long post, but i hope it is bit clearer now what i want to do.

Thanks a lot!

Regards, 
Oliver

**Comment by memsharded** [comment]
> conan export user/stable -s compiler=armgcc -s compiler.version=4.9

This shouldn't work, do you mean ``conan create``?

**Comment by zabeloliver** [comment]
Hi, thanks for your help!
off course it should be `conan create` - was to late for me ...
I set up a few repos where we have a common base to work on:
https://github.com/zabeloliver/conan-sample-project
I tried it localy with conan 0.27.0 and it worked.
I'am in the Slack channel until around 15 o'clock to chat about it!

Thank you very very much for your! I hope at the end we can use this as example for others, since i saw a few requests about this topic at the conan issues.
